operator/controllers/csiscaleoperator_controller.go

```diff
@@ -0,0 +1,1248 @@
+			requests = append(requests, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      CSIScale.Name,
+					Namespace: CSIScale.Namespace,
+				},
+			})
+		}
+		return requests
+	}
+
+	CSIDaemonListFunc := func() []appsv1.DaemonSet {
+		var CSIDaemonSets = []appsv1.DaemonSet{}
+		var DaemonSets = appsv1.DaemonSetList{}
+		_ = mgr.GetClient().List(context.TODO(), &DaemonSets)
+
+		for _, DaemonSet := range DaemonSets.Items {
+			if DaemonSet.Labels[config.LabelProduct] == config.Product {
+				CSIDaemonSets = append(CSIDaemonSets, DaemonSet)
+			}
+		}
+		return CSIDaemonSets
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1.CSIScaleOperator{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(&source.Kind{Type: &corev1.Secret{}},
+			handler.Funcs{
+				CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+					for _, request := range CSIReconcileRequestFunc() {
+						q.Add(request)
+					}
+				},
+				UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+					// TODO: Check if data in e.ObjectNew and e.ObjectOld are same or modified.
+					// Daemon set should update only when data in secret is modified.
+					// Daemon set should not update when secret is updated but not the data.
+					// e.g Secret type, resource version etc is modified.
+					// TODO: Update only those daemon set which are in the same namespace as the secret that triggered the event.
+					for _, daemonSet := range CSIDaemonListFunc() {
+						logger.Info("Secrets were modified. Daemon Set will be updated. Restarting node specific pods.")
+						err = r.rolloutRestartNode(&daemonSet)
+						if err != nil {
+							logger.Error(err, "Unable to update daemon set. Please restart node specific pods manually.")
+						} else {
+							daemonSetRestartedKey, daemonSetRestartedValue = r.getRestartedAtAnnotation(daemonSet.Spec.Template.ObjectMeta.Annotations)
+						}
+					}
+
+					for _, request := range CSIReconcileRequestFunc() {
+						q.Add(request)
+					}
+				},
+				DeleteFunc: func(e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+					for _, request := range CSIReconcileRequestFunc() {
+						q.Add(request)
+					}
+				},
+			}, preds).
+		Owns(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+func Contains(list []string, s string) bool {
+	for _, v := range list {
+		if v == s {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *CSIScaleOperatorReconciler) hasFinalizer(instance *csiscaleoperator.CSIScaleOperator) (bool, error) {
+
+	logger := csiLog.WithName("hasFinalizer")
+
+	accessor, finalizerName, err := r.getAccessorAndFinalizerName(instance)
+	if err != nil {
+		logger.Error(err, "no finalizer found")
+		return false, err
+	}
+
+	logger.Info("returning with finalizer", "name", finalizerName)
+	return Contains(accessor.GetFinalizers(), finalizerName), nil
+}
+
+func Remove(list []string, s string) []string {
```

Comment:
if possible, an we improve this function name and put some description about what it is doing?

Reply:
Done